src/aro/Hideset.zig

```diff
@@ -0,0 +1,187 @@
+fn iterator(self: *const Hideset, idx: Index) Iterator {
+    return Iterator{
+        .slice = self.linked_list.slice(),
+        .i = idx,
+    };
+}
+
+pub fn get(self: *const Hideset, loc: Source.Location) Index {
+    return self.map.get(Identifier.fromLocation(loc)) orelse .sentinel;
+}
+
+pub fn put(self: *Hideset, loc: Source.Location, value: Index) !void {
+    try self.map.put(self.comp.gpa, Identifier.fromLocation(loc), value);
+}
+
+fn ensureUnusedCapacity(self: *Hideset, new_size: usize) !void {
+    try self.linked_list.ensureUnusedCapacity(self.comp.gpa, new_size);
+}
+
+/// Creates a one-item list with contents `identifier`
+fn createNodeAssumeCapacity(self: *Hideset, identifier: Identifier) Index {
+    const next_idx = self.linked_list.len;
+    self.linked_list.appendAssumeCapacity(.{ .identifier = identifier });
+    return @enumFromInt(next_idx);
+}
+
+/// Create a new list with `identifier` at the front followed by `tail`
+pub fn prepend(self: *Hideset, loc: Source.Location, tail: Index) !Index {
+    const new_idx = self.linked_list.len;
+    try self.linked_list.append(self.comp.gpa, .{ .identifier = Identifier.fromLocation(loc), .next = tail });
+    return @enumFromInt(new_idx);
+}
+
+/// Copy a, then attach b at the end
+pub fn @"union"(self: *Hideset, a: Index, b: Index) !Index {
+    var cur: Index = .sentinel;
+    var head: Index = b;
+    try self.ensureUnusedCapacity(self.len(a));
+    var it = self.iterator(a);
+    while (it.next()) |identifier| {
+        const new_idx = self.createNodeAssumeCapacity(identifier);
+        if (head == b) {
+            head = new_idx;
+        }
+        if (cur != .sentinel) {
+            self.linked_list.items(.next)[@intFromEnum(cur)] = new_idx;
+        }
+        cur = new_idx;
+    }
+    if (cur != .sentinel) {
+        self.linked_list.items(.next)[@intFromEnum(cur)] = b;
+    }
+    return head;
+}
+
+pub fn contains(self: *const Hideset, list: Index, str: []const u8) bool {
+    var it = self.iterator(list);
+    while (it.next()) |identifier| {
+        if (mem.eql(u8, str, identifier.slice(self.comp))) return true;
+    }
+    return false;
+}
+
+fn len(self: *const Hideset, list: Index) usize {
+    const nexts = self.linked_list.items(.next);
+    var cur = list;
+    var count: usize = 0;
+    while (cur != .sentinel) : (count += 1) {
+        cur = nexts[@intFromEnum(cur)];
+    }
+    return count;
+}
+
+pub fn intersection(self: *Hideset, a: Index, b: Index) !Index {
+    if (a == .sentinel or b == .sentinel) return .sentinel;
+    self.intersection_map.clearRetainingCapacity();
+
+    var cur: Index = .sentinel;
+    var head: Index = .sentinel;
+    var it = self.iterator(a);
+    var a_len: usize = 0;
+    while (it.next()) |identifier| : (a_len += 1) {
+        const str = identifier.slice(self.comp);
+        try self.intersection_map.put(self.comp.gpa, str, {});
```

Comment:
Is the string equality necessary here? I tested using `Hideset.Identifier` as the key and all tests still passed.

Reply:
I think that's right, updated.